### PR TITLE
Replace Perl with JS.

### PR DESCRIPTION
**Pull request by aminroosta**

### CODE REVIEW

File: scripts/compile.js [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    const config = {
+        languages : common.languages,
+        sections  : ['app', 'static'],
+        root_path : common.root_path,
+        add_translations: false,
+    }
+    if (program.branch === 'translations') {
+        config.languages = ['ACH'];
+    }
+    config.dist_path = Path.join(config.root_path, 'dist');
```
Comment (binary-hamedanchi) [inline scripts/compile.js]: `branch` might be missing here.

File: scripts/compile.js [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        throw new TypeError(`Invalid url ${url}`);
+    }
+}
+
+const createBuilder = async () => {
+    const config = getConfig();
+
+    let static_hash = Math.random().toString(36).substring(2, 10);
+    try {
+        static_hash = await common.readFile(Path.join(config.dist_path, 'version'));
```
Comment (binary-hamedanchi) [inline scripts/compile.js]: Does the perl code uses the current hash only if `--path` is available?

File: scripts/compile.js [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+function create_directories() {
+    const config = getConfig();
+    const mkdir = path => fs.existsSync(path) || fs.mkdirSync(path);
+
+    console.log(`Target: ${config.dist_path}`.yellow);
+    mkdir(Path.join(config.dist_path));
+    config.languages.forEach(lang => {
+        mkdir(Path.join(config.dist_path, lang));
```
Comment (binary-hamedanchi) [inline scripts/compile.js]: `lang` needs to be converted to lower case or better to convert the whole `path` in `mkdir()` above.

File: scripts/compile.js (hunk 1 above)
Comment (aminroosta) [reply scripts/compile.js]: When `branch` is missing then `config.languages` will remain as `common.languages` which has the list of all languages.

File: scripts/compile.js (hunk 2 above)
Comment (aminroosta) [reply scripts/compile.js]: Perl uses current hash only if `--path` is given.

Fixed.

File: scripts/compile.js (hunk 3 above)
Comment (aminroosta) [reply scripts/compile.js]: Fixed.

File: scripts/compile.js (hunk 1 above)
Comment (binary-hamedanchi) [reply scripts/compile.js]: If `branch` is given, does it make the directories as `dist/[branch]/[lang]/...`?

File: scripts/compile.js [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    });
+    await Promise.all(tasks);
+    console.timeEnd(page.save_as);
+}
+
+createDirectories();
+(async () => {
+    const config = getConfig();
+    try {
+        await compile(common.pages.find(p => p.save_as === (program.path || 'home')));
```
Comment (binary-hamedanchi) [inline scripts/compile.js]: Would be more convenient to allow using regular expressions like the current one:
```js
await compile(common.pages.filter(p => new RegExp(program.path, 'i').test(p.save_as)));
```

File: src/templates/_common/components/elements.jsx [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    className,
+    id,
+    href,
+    param = '',
+    target,
+    text,
+    header,
+    p,
+}) => {
+    const content = p ? <p> {text} </p> : text;
```
Comment (negarn) [inline src/templates/_common/components/elements.jsx]: we should remove the white space between `<p>` and `{text}` (?)

File: src/templates/_common/components/elements.jsx [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+};
+
+export const Select = ({
+    id,
+    className,
+    options,
+}) => (
+    <select id={id} className={className} >
+        {options.map(option => (
+            <option value={option.value || undefined} selected={!!option.selected} >
```
Comment (negarn) [inline src/templates/_common/components/elements.jsx]: do we need `|| undefined`?

File: src/templates/_common/components/elements.jsx (hunk 5 above)
Comment (aminroosta) [reply src/templates/_common/components/elements.jsx]: @negar-binary Yea, it's a styling improvement ...however the functionality will remain intact.

File: src/templates/_common/components/elements.jsx (hunk 6 above)
Comment (aminroosta) [reply src/templates/_common/components/elements.jsx]: Yes absolutely :-)
React will omit to render properties with null or undefined values.
We are taking advantages.

Comment (negarn) [reply src/templates/_common/components/elements.jsx]: but if it doesn't exist won't it be undefined by default?

File: src/templates/_common/components/elements.jsx (hunk 5 above)
Comment (negarn) [reply src/templates/_common/components/elements.jsx]: are you sure? as we tested with @binary-hamedanchi it keeps the space in the generated template

Comment (aminroosta) [reply src/templates/_common/components/elements.jsx]: @negarn Does the browser care about the whitespace?

Comment (negarn) [reply src/templates/_common/components/elements.jsx]: I think sometimes, yes, but if we want to rely on it handling most cases then you can leave as is

File: src/templates/_common/components/elements.jsx (hunk 6 above)
Comment (aminroosta) [reply src/templates/_common/components/elements.jsx]: @negar-binary True, you are right. I guess for this it needed to omit the attribute when `option.value === ""`.

**Pull request merged**